Extension/src/LanguageServer/Providers/documentFormattingEditProvider.ts

```diff
@@ -13,53 +13,50 @@ export class DocumentFormattingEditProvider implements vscode.DocumentFormatting
         this.client = client;
     }
 
-    public provideDocumentFormattingEdits(document: vscode.TextDocument, options: vscode.FormattingOptions, token: vscode.CancellationToken): Promise<vscode.TextEdit[]> {
-        return new Promise<vscode.TextEdit[]>((resolve, reject) => {
-            this.client.notifyWhenReady(() => {
-                const filePath: string = document.uri.fsPath;
-                const configCallBack = (editorConfigSettings: any | undefined) => {
-                    const params: FormatParams = {
-                        settings: { ...editorConfigSettings },
-                        uri: document.uri.toString(),
-                        insertSpaces: options.insertSpaces,
-                        tabSize: options.tabSize,
-                        character: "",
-                        range: {
-                            start: {
-                                character: 0,
-                                line: 0
-                            },
-                            end: {
-                                character: 0,
-                                line: 0
-                            }
+    public async provideDocumentFormattingEdits(document: vscode.TextDocument, options: vscode.FormattingOptions, token: vscode.CancellationToken): Promise<vscode.TextEdit[]> {
```

Comment:
`async` is not needed here because the function it decorates does not `await` anything.  It just returns the Promise created by another function.

Reply:
This applies to all the "provide" functions in the Providers folder except `provideWorkspaceSymbols`

Reply:
I take it back.  It looks like notifyWhenReady returns `Promise<Promise<>>` so the async is needed to strip the outer Promise.

That said, we talked offline and since `Promise<Promise<>>` is an extremely odd concept and we're not sure how it will behave in practice, let's go an alternate route here and not allow `async` callbacks to be passed in here.  Instead, let's do
```
  await this.client.notifyWhenReady(() => {});
```
and then put the body of the callback after that await.

I think in the future, a better pattern would be to eliminate callbacks as input to `notifyWhenReady` and require all callers to await its return and perhaps have it provide the `languageClient` as a return value since that is the whole purpose of this function to begin with (to protect access to `this.client.languageClient` before it is ready)

Reply:
please check the modified code.
There were other usages of `notifyWhenReady` in client.ts that  Ididn't change.

Reply:
I will. Yeah, we only need to update the `notifyWhenReady` calls that pass in an `async` function as the argument.